kernel/src/dynamic_process_loading.rs

```diff
@@ -0,0 +1,991 @@
+        command: NonvolatileCommand,
+        offset: usize,
+        length: usize,
+    ) -> Result<usize, ErrorCode> {
+        let mut address: usize = 0;
+
+        let mut state_flag = false; // false if state should be set to AppWrite, true if it should be set to PaddingWrite
+                                    // this is very unelegant but it is a stopgap measure
+
+        self.state.take().map(|state| {
+            match state {
+                State::AppWrite => {
+                    match command {
+                        NonvolatileCommand::UserspaceRead | NonvolatileCommand::UserspaceWrite => {
+                            // need to account for overflow
+                            if offset >= self.new_app_start_addr.get()
+                                || length > self.new_app_length.get()
+                                || offset + length > self.new_app_length.get()
+                            {
+                                // this means the app is out of bounds
+                                return Err(ErrorCode::INVAL);
+                            }
+                        }
+                    }
+                    address = offset + self.new_app_start_addr.get();
+                    state_flag = true;
+                }
+                State::PaddingWrite => {
+                    address = offset;
+                    state_flag = false;
+                }
+                State::Setup => {
+                    // Ideally we should never come here. We aren't supposed to be able to write unless we are in one of the two write states
+                    unimplemented!();
+                }
+                State::Load => {
+                    // Ideally we should never come here. We aren't supposed to be able to write unless we are in one of the two write states
+                    unimplemented!();
+                }
+            }
+            Ok(())
+        });
+        if state_flag {
+            self.state.set(State::AppWrite);
+        } else {
+            self.state.set(State::PaddingWrite);
+        }
+        Ok(address)
+    }
+
+    // Compute the physical address where we should write the data and then write it.
+    //
+    // This function calls write_done() when the Nonvolatile Driver invokes the callback.
+    fn write(
+        &self,
+        command: NonvolatileCommand,
+        user_buffer: &'static mut [u8],
+        offset: usize,
+        length: usize,
+    ) -> Result<(), ErrorCode> {
+        let physical_address = match self.compute_bounds_and_offset(command, offset, length) {
+            Ok(address) => address,
+            Err(e) => return Err(e),
+        };
+
+        let active_len = cmp::min(length, user_buffer.len());
+
+        match command {
+            NonvolatileCommand::UserspaceRead => {
+                self.flash_driver
+                    .read(user_buffer, physical_address, active_len) // change this to no support error code?
+            }
```

Comment:
What is this for?

Reply:
Unnecessary. Removed any reference to NonvolatileCommand from the kernel because we only write